File: Python/textbook_track/chapter04/ba4e.py
# Find a Cyclic Peptide with Theoretical Spectrum Matching an Ideal Spectrum
import itertools as it
import operator
import sys

from collections import Counter
from copy import deepcopy

from textbook_track.resources.utils import read_amino_acid_mass_table


class Peptide:
    def __init__(self, masses):
        self._masses = masses
        self._total_mass = sum(masses)

    def __str__(self):
        return '-'.join(map(str, self._masses))

    def __len__(self):
        return len(self._masses)

    @property
    def masses(self):
        return tuple(self._masses)

    @property
    def total_mass(self):
        return self._total_mass

    def update(self, mass):
        self._masses.append(mass)
        self._total_mass += mass


def convert_to_intlist(line):
    return [int(elem) for elem in line.split()]


def calc_peptide_prefix_masses(peptide):
    return list(it.accumulate(peptide.masses, operator.add, initial=0))


def expand(peptides, amino_acid_masses):
    for peptide in peptides:
        for mass in amino_acid_masses:
            new_peptide = deepcopy(peptide)
            new_peptide.update(mass)
            yield new_peptide


def calc_theoretical_linear_spectrum(peptide):
    prefix_masses = calc_peptide_prefix_masses(peptide)
    n = len(peptide)
    spectrum = [0]
    for start_ix in range(n):
        for end_ix in range(start_ix+1, n+1):
            subpeptide_mass = prefix_masses[end_ix] - prefix_masses[start_ix]
            spectrum.append(subpeptide_mass)
    return Counter(spectrum)


def calc_theoretical_cyclic_spectrum(peptide):
    prefix_masses = calc_peptide_prefix_masses(peptide)
    peptide_mass = prefix_masses[-1]
    n = len(peptide)
    cyclospectrum = [0]
    for start_ix in range(n):
        for end_ix in range(start_ix+1, n+1):
            subpeptide_mass = prefix_masses[end_ix] - prefix_masses[start_ix]
            cyclospectrum.append(subpeptide_mass)
            if start_ix > 0 and end_ix < n:
                cyclospectrum.append(peptide_mass - subpeptide_mass)
    return Counter(cyclospectrum)


def is_peptide_consistent_with_experimental_spectrum(peptide, experimental_spectrum):
    linear_spectrum = calc_theoretical_linear_spectrum(peptide)
    return all(experimental_spectrum.get(mass, 0) >= count for mass, count in linear_spectrum.items())


def cyclopeptide_sequencing(experimental_spectrum, mass_table):
    amino_acid_masses = frozenset(mass_table.values())
    parent_mass = max(experimental_spectrum)
    experimental_spectrum = Counter(experimental_spectrum)
    peptides = [Peptide([])]
    results = []
    while peptides:
        candidate_peptides = expand(peptides, amino_acid_masses)
        peptides = []
        for peptide in candidate_peptides:
            if peptide.total_mass == parent_mass:
                cyclospectrum = calc_theoretical_cyclic_spectrum(peptide)
                if cyclospectrum == experimental_spectrum:
                    results.append(peptide)
            elif is_peptide_consistent_with_experimental_spectrum(peptide, experimental_spectrum):
                peptides.append(peptide)
    return results


def main():
    data = sys.stdin.read().splitlines()
    experimental_spectrum = convert_to_intlist(data[0])
    mass_table = read_amino_acid_mass_table()
    result = cyclopeptide_sequencing(experimental_spectrum, mass_table)
    print(' '.join(map(str, result)))


if __name__ == '__main__':
    main()

File: Scala/src/main/scala/TextbookTrack/Chapter04/BA4E.scala
package TextbookTrack.Chapter04

object BA4E {
  import scala.language.implicitConversions
  import scala.annotation.tailrec
  import TextbookTrack.Utils.IntegerMassTable

  final case class Peptide(masses: List[Int]) {
    override def toString: String = masses.mkString("-")

    val length: Int = masses.length
    val totalMass: Int = masses.sum
  }

  private def convertToIntList(line: String): List[Int] = line.split(" ").map(_.toInt).toList

  private implicit def toMassCounts(masses: List[Int]): Map[Int, Int] =
    masses.groupBy(identity).view.mapValues(_.length).toMap

  private def calcPeptidePrefixMasses(peptide: Peptide): Vector[Int] =
    peptide.masses.reverseIterator.scanLeft(0){ _ + _ }.toVector

  private def calcTheoreticalLinearSpectrum(peptide: Peptide): List[Int] = {
    val prefixMasses: Vector[Int] = calcPeptidePrefixMasses(peptide)
    val n: Int = peptide.length
    (0 until n).foldLeft(List(0)){ case (spectrum, startIx) =>
      (startIx + 1 to n).foldLeft(spectrum){ case (acc, endIx) =>
        val subpeptideMass: Int = prefixMasses(endIx) - prefixMasses(startIx)
        subpeptideMass :: acc
      }
    }.sorted
  }

  private def calcTheoreticalCyclicSpectrum(peptide: Peptide): List[Int] = {
    val prefixMasses: Vector[Int] = calcPeptidePrefixMasses(peptide)
    val peptideMass: Int = prefixMasses.last
    val n: Int = peptide.length
    (0 until n).foldLeft(List(0)){ case (cycloSpectrum, startIx) =>
      (startIx + 1 to n).foldLeft(cycloSpectrum){ case (acc, endIx) =>
        val subpeptideMass: Int = prefixMasses(endIx) - prefixMasses(startIx)
        if (startIx > 0 && endIx < n) (peptideMass - subpeptideMass) :: subpeptideMass :: acc
        else subpeptideMass :: acc
      }
    }.sorted
  }

  private def isPeptideConsistentWithExperimentalSpectrum(peptide: Peptide,
                                                          experimentalSpectrum: Map[Int, Int]): Boolean = {
    val linearSpectrum: Map[Int, Int] = calcTheoreticalLinearSpectrum(peptide)
    linearSpectrum.forall{ case (mass, count) => experimentalSpectrum.getOrElse(mass, 0) >= count }
  }

  private def expand(peptides: List[Peptide], aminoAcidMasses: Set[Int]): Iterator[Peptide] =
    peptides.iterator.flatMap{ peptide => aminoAcidMasses.map(mass => peptide.copy(mass :: peptide.masses)) }

  def cyclopeptideSequencing(experimentalSpectrum: Map[Int, Int]): List[Peptide] = {
    val aminoAcidMasses: Set[Int] = IntegerMassTable.values.toSet
    val parentMass: Int = experimentalSpectrum.keysIterator.max

    @tailrec
    def loop(acceptedPeptides: List[Peptide], currentPeptides: List[Peptide]): List[Peptide] = {
      if (currentPeptides.isEmpty) acceptedPeptides
      else {
        val candidatePeptides: Iterator[Peptide] = expand(currentPeptides, aminoAcidMasses)
        val (updatedAcceptedCandidates, candidates): (List[Peptide], List[Peptide]) =
          candidatePeptides.foldLeft((acceptedPeptides, List.empty[Peptide])){
            case ((accepted, candidates), peptide) =>
              if (peptide.totalMass == parentMass) {
                val cycloSpectrum: Map[Int, Int] = calcTheoreticalCyclicSpectrum(peptide)
                if (cycloSpectrum == experimentalSpectrum) (peptide :: accepted, candidates)
                else (accepted, candidates)
              }
              else if (isPeptideConsistentWithExperimentalSpectrum(peptide, experimentalSpectrum))
                (accepted, peptide :: candidates)
              else (accepted, candidates)
          }
        loop(updatedAcceptedCandidates, candidates)
      }
    }

    loop(Nil, List(Peptide(Nil)))
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val experimentalSpectrum: Map[Int, Int] = convertToIntList(reader.next())
    val result: List[Peptide] = cyclopeptideSequencing(experimentalSpectrum)
    println(result.mkString(" "))
  }
}
